share/shwap/range_namespace_data.go

```diff
@@ -0,0 +1,215 @@
+			if !namespace.Equals(shrs[from.Col+offset].Namespace()) {
+				return RangeNamespaceData{},
+					fmt.Errorf("targeted namespace was not found in share at {Row: %d, Col:%d}",
+						row, from.Col+offset,
+					)
+			}
+		}
+
+		proof, err := tree.ProveRange(from.Col, end)
+		if err != nil {
+			return RangeNamespaceData{}, err
+		}
+
+		nsData = append(nsData, RowNamespaceData{Shares: shrs[from.Col:end], Proof: &proof})
+
+		// reset from.Col as we are moving to the next row.
+		from.Col = 0
+		row++
+	}
+	return RangeNamespaceData{nsData}, nil
+}
+
+// Validate performs a validation of the incoming data. It ensures that the response contains proofs and performs
+// data validation in case the user has requested it.
+func (rngdata *RangeNamespaceData) Validate(root *share.AxisRoots, req *RangeNamespaceDataID) error {
+	_, err := rngdata.IsEmpty()
+	if err != nil {
+		return fmt.Errorf("RangeNamespaceData: empty data: %w", err)
+	}
+
+	if rngdata.NamespaceData[0].Proof.Start() != req.ShareIndex {
+		return fmt.Errorf("RangeNamespaceData: invalid start of the range: want: %d, got: %d",
+			req.ShareIndex, rngdata.NamespaceData[0].Proof.Start(),
+		)
+	}
+
+	// verify shares amount
+	rawShares := rngdata.Flatten()
+	odsFromIdx, err := SampleCoordsAs1DIndex(
+		SampleCoords{req.RowIndex, req.ShareIndex}, len(root.RowRoots)/2,
+	)
+	if err != nil {
+		return err
+	}
+
+	odsToIdx, err := SampleCoordsAs1DIndex(req.To, len(root.RowRoots)/2)
+	if err != nil {
+		return err
+	}
+	if len(rawShares) != (odsToIdx - odsFromIdx + 1) { // `+1` because both indexes are inclusive
+		return fmt.Errorf("RangeNamespaceData: shares amount mismatch: want: %d, got: %d",
+			odsToIdx-odsFromIdx+1, len(rawShares),
+		)
+	}
+
+	rowStart := req.RowIndex
+	for i, row := range rngdata.NamespaceData {
+		verified := row.Proof.VerifyInclusion(
+			share.NewSHA256Hasher(),
+			req.DataNamespace.Bytes(),
+			libshare.ToBytes(row.Shares),
+			root.RowRoots[rowStart+i],
+		)
+		if !verified {
+			return fmt.Errorf("RangeNamespaceData: %w at row: %d", ErrFailedVerification, rowStart+i)
+		}
+	}
+	return nil
+}
+
+// IsEmpty verifies whether the underlying `NamespaceData` is empty.
+// It returns an error in case the `RangeNamespaceData` is completely empty
+func (rngdata RangeNamespaceData) IsEmpty() (bool, error) {
+	if len(rngdata.NamespaceData) == 0 {
+		return true, errors.New("namespace data is empty")
+	}
+
+	for _, row := range rngdata.NamespaceData {
+		if row.Proof == nil {
+			return true, errors.New("proof list for the row is empty")
+		}
+		if row.IsEmpty() {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (rngdata *RangeNamespaceData) ToProto() *pb.NamespaceData {
+	return rngdata.NamespaceData.ToProto()
+}
+
+// ProveRange proves that a range of shares exist in the set of rows that are part of the merkle tree of the data root.
+func (rngdata *RangeNamespaceData) ProveRange(roots *share.AxisRoots, startRow int) *types.ShareProof {
```

Comment:
what is this proving? how do you specify the set of shares to be proved?

Reply:
RangeNamespaceData is a slice of the RowNamespaceData, which contains shares and nmt proofs for these shares:

```
type RowNamespaceData struct {
	Shares []libshare.Share `json:"shares"` // Shares within the namespace.
	Proof  *nmt.Proof       `json:"proof"`  // Proof of the shares' inclusion in the namespace.
}
```
ProveRange builds a data root proof of the shares.